### PR TITLE
fix #350

### LINKED ISSUE #350: appstream-compose mangles values between some tags
We've noticed this some time ago in Flathub but I don't think anyone reported a bug here. Apparently some tags like `<strong>` and `<em>` end up as `(null)` in the resulting app-info file. Some examples:

Midori (https://github.com/midori-browser/core/blob/master/data/org.midori_browser.Midori.appdata.xml.in#L25):

```
<description><p>Midori is a lightweight yet powerful web browser which runs just as well on little embedded computers named for delicious pastries as it does on beefy machines with a core temperature exceeding that of planet earth. And it looks good doing that, too. Oh, and of course it&apos;s free software.</p><p>(null)</p><ul><li>Adblock filter list support.</li><li>Private browsing.</li><li>Manage cookies and scripts.</li></ul><p>(null)</p><ul><li>Open a 1000 tabs instantly.</li><li>Easy web apps creation.</li><li>Customizable side panels.</li><li>User scripts and styles a la Greasemonkey.</li><li>Web developer tools powered by WebKit.</li></ul></description>
```

Cheese (https://github.com/flathub/org.gnome.Cheese/blob/master/org.gnome.Cheese.appdata.xml.in.patch)

```
<releases>
      <release timestamp="1568073600" version="3.34.0">
        <url type="unknown">https://gitlab.gnome.org/GNOME/cheese/raw/3.34.0/NEWS</url>
        <description><p>(null)</p></description>
      </release>
    </releases>
```

Firefox (appdata at https://paste.xinu.at/9B4HNN4/):

```
<description><p>(null)</p><p>Meet Firefox Quantum. Fast for good.</p><p>Features:</p><ul><li>A powerful, new engine that’s built for rapidfire performance.</li><li>Better, faster page loading that uses less computer memory.</li><li>Gorgeous design and smart features for intelligent browsing.</li></ul><p>(null)</p><p>Crazy powerful browser engine? Check. Less time waiting around for pages to load? Also, check. Firefox Quantum is twice as fast as Firefox was before.</p><p>(null)</p><p>Whether you’ve opened 10 or 1000, switching between tabs is faster than ever with Firefox’s new, responsive engine.</p><p>(null)</p><p>The most powerful private browsing mode with added tracking protection.</p><ul><li>(null)</li><li>(null)</li><li>(null)</li></ul><p>(null)</p><ul><li>(null)</li><li>(null)</li><li>(null)</li></ul><p>(null)</p><ul><li>Seamlessly access passwords, bookmarks and more. Plus, use our Send Tabs feature to instantly share open tabs between desktop, mobile and tablet.</li></ul><p>(null)</p><p>Firefox is made by Mozilla, the non-profit champions of a healthy internet. Mozilla also tackles issues like privacy, misinformation and trolling by investing in fellowships, campaigns and new technologies designed to make the internet healthier.</p></description>
```

### PR DESCRIPTION
**Pull request by guinux**
Add support for \<em> and \<code> tags, as described in https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-description .

### CODE REVIEW

File: libappstream-glib/as-node.c [hunk 1]
```diff
@@ -555,6 +555,8 @@ typedef struct {
 	AsNode			*current;
 	AsNodeFromXmlFlags	 flags;
 	const gchar * const	*locales;
+	guint8 is_em_text;
+	guint8 is_code_text;
```
Comment (hughsie) [inline libappstream-glib/as-node.c:559]: spacing weird

File: libappstream-glib/as-node.c [hunk 2]
```diff
@@ -604,6 +606,16 @@ as_node_start_element_cb (GMarkupParseContext *context,
 	AsNode *current;
 	guint i;
 
+	/* do not create a child node for em and code tags */
+	if (g_strcmp0 (element_name, "em") == 0) {
+		helper->is_em_text = 1;
+		return;
+	} else if (g_strcmp0 (element_name, "code") == 0) {
```
Comment (hughsie) [inline libappstream-glib/as-node.c:613]: no need for else if when returning.

File: libappstream-glib/as-node.c [hunk 3]
```diff
@@ -662,6 +674,15 @@ as_node_end_element_cb (GMarkupParseContext *context,
 			GError             **error)
 {
 	AsNodeToXmlHelper *helper = (AsNodeToXmlHelper *) user_data;
+
+	/* do not create a child node for em and code tags */
+	if (g_strcmp0 (element_name, "em") == 0) {
+		helper->is_em_text = 0;
+		return;
+	} else if (g_strcmp0 (element_name, "code") == 0) {
```
Comment (hughsie) [inline libappstream-glib/as-node.c:682]: same

**Comment by hughsie** [comment]
This also needs a ton of self tests to be added too, please.

**Comment by nbenitez** [comment]
Should this MR be closed as per commit 674490bd54ff206f213ca4547db7fdb591a0fb3d ?

**Comment by hughsie** [comment]
I think so; thanks all.

**Pull request closed without merging**